### PR TITLE
Feature/add edit user release permissions

### PR DESCRIPTION
**Pull request by jamescutts**
This PR adds functionality for adding / removing release roles from users.

image

The UI has been updated to save changes as they occur rather than submitting all changes at once.

### Logic
- Any release in the system can have a role assigned regardless of its state (Draft etc.)
- Any role can be added to a user for a release
- Adding a role a user already has on a specific release should result in an error being displayed.
- Users will receive an email informing them of their updated access.

### Refactoring
This quite extensively refactors the controllers for user management, splitting them out into their respective functions (users, pre-release and invites) and adjusts their paths to be under a common `user-management` route.

It also fleshes out some of the controllers with full methods for CRUD actions even though these might not currently all be used.

This should allows easier maintenance in the future and will also facilitate more granular permissions should these be required as controllers are now grouped by logic area.

### Suggested improvements
The following improvements are suggested but not covered as part of this task and may be raised separately.

- Prevent a user editing themselves
- Creation of specific user management roles/claims so that BAU is not required.
- Add messaging if the BAU role is selected saying release roles do not need to be assigned and disable that section of the UI.

### CODE REVIEW

**Comment by jamescutts** [comment]
Just spotted i need to swap out the endpoint it get releases from to a new one that includes everything

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx [hunk 1]
```diff
@@ -91,71 +96,122 @@ const UserInvitePage = ({
...
+                    label="Role"
+                    name="selectedRoleId"
+                    options={model?.roles.map(role => ({
+                      label: role.name,
+                      value: role.id,
+                    }))}
+                  />
+                </FormFieldset>
+
+                {/* <FormFieldset
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:147]: Why is this all commented out?

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx [hunk 2]
```diff
@@ -91,71 +96,122 @@ const UserInvitePage = ({
...
+                  id={`${formId}-selectedRoleIdFieldset`}
+                  legend="Role"
+                  legendSize="m"
+                  hint="The users role within the service."
+                >
+                  <FormFieldSelect
+                    id={`${formId}-selectedRoleId`}
+                    label="Role"
+                    name="selectedRoleId"
+                    options={model?.roles.map(role => ({
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:140]: Would add a `placeholder` like 'Choose role'

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx [hunk 3]
```diff
@@ -91,71 +96,122 @@ const UserInvitePage = ({
...
-                      legendSize="m"
-                      hint="Most users should be added with the analyst role"
-                    >
+      {model && (
+        <Formik<FormValues>
+          enableReinitialize
+          initialValues={{
+            userEmail: '',
+            selectedRoleId:
+              orderBy(model.roles, role => role.name)?.[0]?.id ?? '',
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:105]: Should we be defaulting to the first role? I would recommend leaving this empty to make sure the user is aware of what they're selecting

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx (hunk 1 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:147]: Part of  EES-622 which i was doing at the same time however they diverged a bit.

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx (hunk 3 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:105]: Yeah the first role is the most applicable (we only have two) 99% of the time a user will be added as an analyst which is what this defaults too. Saves the user an extra step.

File: src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx (hunk 2 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/users/UserInvitePage.tsx:140]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public class PrereleaseUsersController : ControllerBase
+    {
+        private readonly IUserManagementService _userManagementService;
+
+        public PrereleaseUsersController(IUserManagementService userManagementService)
+        {
+            _userManagementService = userManagementService;
+        }
+
+        [HttpGet("user-management/pre-release")]
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs:23]: Feels a bit wordy to be prefixing everything with `user-management`. Why don't we simply name it `users`?

Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs:23]: Was aligning this with the namespaces and something common that describes what these split controllers do. Did have users earlier on but it didn't really work as a nice base given theres /users path so i added -management.

Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs:23]: Wouldn't say there's any reason why controllers that share the same namespace in code need to share the same prefix in the API. I think the controllers being split up is good stuff, but tacking on the prefix just feels unnecessarily verbose and is a little bit anti-REST.

From what I can see, the endpoints could just be:
- `/users`
- `/pre-release-users` (or even better - `/api/users?type=pre-release`
- `/invites`
- `/roles`
- `/release-roles`
- `/releases` - isn't there already a releases endpoint for the rest of admin???

Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs:23]: I'm not too concerned by the prefix so not intending to change it.

Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/PrereleaseUsersController.cs:23]: The logic and contents of new new release endpoint are quite specific to user management where a simplified fill list is required regardless of any access permissions. The exact need implementation of this will probably change when the next user management and permissions epic starts.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/BAU/BauUsersControllerPermissionsTests.cs [hunk 5]
```diff
@@ -1,5 +1,4 @@
-using GovUk.Education.ExploreEducationStatistics.Admin.Controllers.Api.BAU;
-using GovUk.Education.ExploreEducationStatistics.Admin.Controllers.Api.BAU.UserManagement;
+using GovUk.Education.ExploreEducationStatistics.Admin.Controllers.Api.UserManagement;
 using GovUk.Education.ExploreEducationStatistics.Admin.Security;
 using Xunit;
 using static GovUk.Education.ExploreEducationStatistics.Admin.Tests.Services.PermissionTestUtil;
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/BAU/BauUsersControllerPermissionsTests.cs:4]: I think this class may need renaming now from `BauUsersControllerPermissionsTests` to something else.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/UserInviteSubmission.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+namespace GovUk.Education.ExploreEducationStatistics.Admin.ViewModels
+{
+    public class UserInviteSubmission
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/UserInviteSubmission.cs:3]: I think everywhere else we are using Request for the naming so this would be `UserInviteRequest`

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/UsersController.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+            return NotFound();
+        }
+
+        [HttpGet("user-management/users/{userId}")]
+        [ProducesResponseType(200)]
+        [ProducesResponseType(404)]
+        public async Task<ActionResult<UserViewModel>> GetUser(string userId)
+        {
+            var user = await _userManagementService.GetAsync(userId);
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/UserManagement/UsersController.cs:49]: I would have changed the `_userManagementService.GetAsync` to use `IPersistenceHelper#CheckEntityExists` to get the user and change the signature of method `GetAsync` to return `Task<Either<ActionResult, UserViewModel>>`  because then this method could all be simplified to `return await _userManagementService.GetAsync(userId).HandleFailuresOrOk()`. The PersistenceHelper returns NotFound at the Service level.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs [hunk 8]
```diff
@@ -54,6 +62,75 @@ public async Task<List<UserViewModel>> ListAsync()
...
+                            r.Id == userReleaseRoleId && r.UserId == userId);
+                    _contentDbContext.Remove(entityToRemove);
+                    await _contentDbContext.SaveChangesAsync();
+                })
+                .OnSuccess(_ => true);
+        }
+
+        public async Task<List<IdTitlePair>> ListReleasesAsync()
+        {
+            return await _contentDbContext.Releases.Select(r => new IdTitlePair
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:127]: I think you should `.Include` publication here. This wouldn't work called on its own unless something else has already loaded the Publication into the Context.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs [hunk 9]
```diff
@@ -239,5 +317,37 @@ private void SendInviteEmail(string email)
...
+
+        private void SendNewReleaseRoleEmail(Guid userId, IdTitlePair publication, IdTitlePair release, EnumExtensions.EnumValue role)
+        {
+            var uri = _configuration.GetValue<string>("AdminUri");
+            var template = _configuration.GetValue<string>("NotifyReleaseRoleTemplateId");
+            var email = _usersAndRolesDbContext.Users.FirstOrDefault(x => x.Id == userId.ToString())?.Email;
+
+            var emailValues = new Dictionary<string, dynamic>
+            {
+                {"url", "https://" + uri + "/publication/" + publication.Id + "/release/" + release.Id + "/summary"},
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:329]: `$"https://{uri}/publication/{publication.Id}/release/{release.Id}/summary"` might be nicer to read?

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs [hunk 10]
```diff
@@ -239,5 +317,37 @@ private void SendInviteEmail(string email)
...
+        }
+
+        private async Task<Either<ActionResult, bool>> CheckIfUserAlreadyHasReleaseRole(Guid userId,
+            UserReleaseRoleSubmission userReleaseRole)
+        {
+            var existing = await _contentDbContext.UserReleaseRoles.FirstOrDefaultAsync(r =>
+                r.UserId == userId && r.ReleaseId == userReleaseRole.ReleaseId &&
+                r.Role == userReleaseRole.ReleaseRole);
+
+            if (existing == null)
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:345]: This doesn't look right. The method is called `CheckIfUserAlreadyHasReleaseRole`, so should it be `if (existing != null) { return ValidationActionResult(UserAlreadyHasReleaseRole } return false` ?

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/UserViewModel.cs [hunk 11]
```diff
@@ -16,11 +16,17 @@ public class UserViewModel
         
         public string Role { get; set; }
         
-        public List<UserReleaseRole> UserReleaseRoles { get; set; }
+        public List<UserReleaseRoleViewModel> UserReleaseRoles { get; set; }
     }
 
-    public class UserReleaseRole
+    public class UserReleaseRoleSubmission
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/UserViewModel.cs:22]: I'm not a big fan of this being in the same file and I would call it `Request` rather than `Submission` like we are using `Request` everywhere else.

Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/UserViewModel.cs:22]: Ah cool, not a problem

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs (hunk 9 above)
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:329]: Ah, i thought i had done that already

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs (hunk 8 above)
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:127]: I've not encountered that error but it probably wouldn't hurt, i think in theory if your using select it executes and returns that specific sql command, i think it would break if you used orderby after the select on something that comes from the publication.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs (hunk 10 above)
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserManagementService.cs:345]: Case of bad naming i think, the logic does work
